### DISCUSSION TITLE
New "dependency" keyword

### DISCUSSION THREAD

**Discussion by ekisielinski** [dotnet/csharplang#3886] [General] [answered]
Instead of this:

    public class MyService
    {
        IServiceA serviceA;
        readonly IServiceB serviceB;
        IServiceC? serviceC;

        public MyService( IServiceA serviceA, IServiceB serviceB, IServiceC? serviceC )
        {
            if (serviceA is null) thrown new ArgumentNullException(...);
            this.serviceA = serviceA;        
            if (serviceB is null) thrown new ArgumentNullException(...);
            this.serviceB = serviceB;        
            this.serviceC = serviceC; // nullable, no checking for null
        }
    }

We could write this:

    public class MyService
    {
        dependency IServiceA serviceA;
        dependency readonly IServiceB serviceB;
        dependency IServiceC? serviceC;
    }

Constructor would be created by compiler with dependencie with the same order as declared.

Additionaly we could create method:

    private void ValidateDependencies()
    {
         // additional checking
    }

Which would be called in auto-generated constructor.

:)

**Comment by HaloFour** [comment, marked as answer]
With records you will be able to do this:

```cs
public record MyService(IServiceA serviceA, IServiceB serviceB, IServiceC serviceC) {
    // members go here
}
```

It's expected that primary constructors (the feature that allows for a parameter list to be specified after the type name) will be supported for normal classes by C# 10.0 complete with a constructor block where you can perform validation:

```cs
public class MyService(IServiceA serviceA, IServiceB serviceB, IServiceC serviceC) {
    public MyService {
        if (serviceA is null) throw new ArgumentNullException(...);
        if (serviceB is null) throw new ArgumentNullException(...);
    }

    // members go here
}
```

With https://github.com/dotnet/csharplang/issues/2145 you could possibly also write it as:

```cs
public class MyService(IServiceA serviceA!!, IServiceB serviceB!!, IServiceC serviceC) {
    // members go here
}
```

**Comment by PathogenDavid** [comment]
This feels like a very domain-specific problem that would be better solved with a domain-specific source generator.

**Reply by Richiban** [reply]
It's definitely not domain-specific, although it might be coding style-specific.
